### PR TITLE
Dev-datamigration-Microsoft.DataMigration-2022-03-30-preview

### PR DESCRIPTION
**Pull request by ashutoshsuman99**
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i>

### Changelog
Add a changelog entry for this PR by answering the following questions:
  1. What's the purpose of the update?
      - [ ] new service onboarding
      - [x] new API version
      - [ ] update existing version for new feature
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify
  2. When are you targeting to deploy the new service/feature to public regions? Please provide the date or, if the date is not yet available, the month.
  3. When do you expect to publish the swagger? Please provide date or, the the date is not yet available, the month.
  4. If updating an existing version, please select the specific language SDKs and CLIs that must be refreshed after the swagger is published.
      - [ ] SDK of .NET (need service team to ensure code readiness)
      - [ ] SDK of Python
      - [ ] SDK of Java
      - [ ] SDK of Js
      - [ ] SDK of Go
      - [ ] PowerShell
      - [ ] CLI
      - [ ] Terraform
      - [ ] No refresh required for updates in this PR

### Contribution checklist:
- [x] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of "no breaking changes"
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist 

> **Applicability**: :warning: 
>
> If your changes encompass only the following scenarios, you should SKIP this section, as these scenarios do not require ARM review.
> - Change to data plane APIs 
> - Adding new properties 
> - All removals

Otherwise your PR may be subject to ARM review requirements. Complete the following:
- [x] Check this box if any of the following apply to the PR so that label "WaitForARMFeedback" will be added automatically to begin ARM API Review. Failure to comply may result in delays to the manifest.
  - Adding a new service
  - Adding new API(s)
  - Adding a new API version
    -[ ] To review changes efficiently, ensure you are using OpenAPIHub to initialize the PR for adding a new version. More details, refer to the [wiki](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/208/OpenAPI-Hub-Adding-new-API-version).

- [ ] Ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If any of the following scenarios apply to the PR, request approval from the Breaking Change Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 
- [ ] Removing API(s) in a stable version
- [ ] Removing properties in a stable version
- [ ] Removing API version(s) in a stable version
- [ ] Updating API in a stable or public preview version with Breaking Change Validation errors
- [ ] Updating API(s) in public preview over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Addition details on the process and office hours are on the [Breaking change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

### CODE REVIEW

**Comment by qiaozha** [comment]
@raych1 Can you take a look at this model validation error ? seems the discriminator value has already been added in the taskType definition. Is this a bug for model validation ?

**Comment by raosuhas** [comment]
>     "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Sql/servers/{sqlDbInstanceName}/providers/Microsoft.DataMigration/databaseMigrations/{targetDbName}": {

rename to migration name. Assuming this does not have to match the targetdb name ?

---
Refers to: specification/datamigration/resource-manager/Microsoft.DataMigration/preview/2022-03-30-preview/sqlmigration.json:19 in 7eb26c8. [](commit_id = 7eb26c84f50274e95658c4dfc02ac0a15c40f78b, deletion_comment = False)

**Comment by raosuhas** [comment]
>             "name": "migrationOperationId",

So I assume everytime you create this databaseMigrations resource it starts a new Migrationoperation and then. It might have been better to have a separate resource that defines the migration and one to define the migration configuration . so you could have a history of operations etc and they are 2 separate things. But I see that you have already implemented this in this fashion for the managedinstance .. So it is fine if you want to keep this consistent but consider if this is something you would want in the future. 

---
Refers to: specification/datamigration/resource-manager/Microsoft.DataMigration/preview/2022-03-30-preview/sqlmigration.json:40 in 7eb26c8. [](commit_id = 7eb26c84f50274e95658c4dfc02ac0a15c40f78b, deletion_comment = False)

**Comment by raosuhas** [comment]
>         },

Why is the scope duplicated here ? It seems like since this is an extension resource the scope is defined by the resource itself ? 

---
Refers to: specification/datamigration/resource-manager/Microsoft.DataMigration/preview/2022-03-30-preview/sqlmigration.json:1262 in 7eb26c8. [](commit_id = 7eb26c84f50274e95658c4dfc02ac0a15c40f78b, deletion_comment = False)

**Comment by raosuhas** [comment]
>         },

What happens if the user provides a different scope here from the uri scope ?

---
In reply to: [1068534088](https://github.com/Azure/azure-rest-api-specs/pull/18213#issuecomment-1068534088) [](http://example.com/codeflow?ancestors=1068534088)

---
Refers to: specification/datamigration/resource-manager/Microsoft.DataMigration/preview/2022-03-30-preview/sqlmigration.json:1262 in dad3b90. [](commit_id = dad3b90cf442206537fe0c525221c476a6cc520c, deletion_comment = False)

File: specification/datamigration/resource-manager/Microsoft.DataMigration/preview/2022-03-30-preview/examples/SqlDbCreateOrUpdateDatabaseMigrationMAX.json [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+          "trustServerCertificate": true,
+          "authentication": "WindowsAuthentication"
+        },
+        "sqlDataCopyThresholds": {
+          "cidxRowThreshold": 10000,
+          "cidxKbsThreshold": 2000
+        },
+        "tableList": [
+          "[Schema1].[TableName1]",
+          "[Schema2].[TableName2]"
```
Comment (raosuhas) [inline specification/datamigration/resource-manager/Microsoft.DataMigration/preview/2022-03-30-preview/examples/SqlDbCreateOrUpdateDatabaseMigrationMAX.json:36]: The problem here is that this violates the GET - PUT parity for this resource. You PUT something and the GET returns something else. This is not how REST is supposed to work.

**Comment by raosuhas** [comment]
>             "description": "The child resources to include in the response.",

Update description.

---
Refers to: specification/datamigration/resource-manager/Microsoft.DataMigration/preview/2022-03-30-preview/sqlmigration.json:50 in 7eb26c8. [](commit_id = 7eb26c84f50274e95658c4dfc02ac0a15c40f78b, deletion_comment = False)

**Comment by ashutoshsuman99** [comment]
> > ```
> >     },
> > ```
> 
> Why is the scope duplicated here ? It seems like since this is an extension resource the scope is defined by the resource itself ?
> 
> Refers to: specification/datamigration/resource-manager/Microsoft.DataMigration/preview/2022-03-30-preview/sqlmigration.json:1262 in [7eb26c8](https://github.com/Azure/azure-rest-api-specs/commit/7eb26c84f50274e95658c4dfc02ac0a15c40f78b). [](commit_id = [7eb26c8](https://github.com/Azure/azure-rest-api-specs/commit/7eb26c84f50274e95658c4dfc02ac0a15c40f78b), deletion_comment = False)

Yes, this is a proxy resource. Scope in the request body is to validate permission. When you send request, arm first validates that user has permission on MI/VM/SQLDB. Arm can validate permission on property in input, so we have specified 'scope' in manifest as the property to be validated.

File: specification/datamigration/resource-manager/Microsoft.DataMigration/preview/2022-03-30-preview/examples/SqlDbCreateOrUpdateDatabaseMigrationMAX.json (hunk 1 above)
Comment (ashutoshsuman99) [reply specification/datamigration/resource-manager/Microsoft.DataMigration/preview/2022-03-30-preview/examples/SqlDbCreateOrUpdateDatabaseMigrationMAX.json:36]: We do return the tables that are being migrated, but that are part of property in response. We will add tableList as well to the response.

**Comment by ashutoshsuman99** [comment]
> > ```
> >     },
> > ```
> 
> What happens if the user provides a different scope here from the uri scope ?
> 
> In reply to: [1068534088](https://github.com/Azure/azure-rest-api-specs/pull/18213#issuecomment-1068534088)
> 
> Refers to: specification/datamigration/resource-manager/Microsoft.DataMigration/preview/2022-03-30-preview/sqlmigration.json:1262 in [dad3b90](https://github.com/Azure/azure-rest-api-specs/commit/dad3b90cf442206537fe0c525221c476a6cc520c). [](commit_id = [dad3b90](https://github.com/Azure/azure-rest-api-specs/commit/dad3b90cf442206537fe0c525221c476a6cc520c), deletion_comment = False)

We return an error in that case. Invalid value provided for parameter 'Scope'.

**Comment by ashutoshsuman99** [comment]
> > ```
> > "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Sql/servers/{sqlDbInstanceName}/providers/Microsoft.DataMigration/databaseMigrations/{targetDbName}": {
> > ```
> 
> rename to migration name. Assuming this does not have to match the targetdb name ?
> 
> Refers to: specification/datamigration/resource-manager/Microsoft.DataMigration/preview/2022-03-30-preview/sqlmigration.json:19 in [7eb26c8](https://github.com/Azure/azure-rest-api-specs/commit/7eb26c84f50274e95658c4dfc02ac0a15c40f78b). [](commit_id = [7eb26c8](https://github.com/Azure/azure-rest-api-specs/commit/7eb26c84f50274e95658c4dfc02ac0a15c40f78b), deletion_comment = False)

This operation actually creates a targetDb with the given name in target platform like SQL MI, SQL VM and SQL DB. That is why we have added targetDbName to the parameter to be more clear.

**Comment by MYethirajam** [comment]
> > ```
> >         "name": "migrationOperationId",
> > ```
> 
> So I assume everytime you create this databaseMigrations resource it starts a new Migrationoperation and then. It might have been better to have a separate resource that defines the migration and one to define the migration configuration . so you could have a history of operations etc and they are 2 separate things. But I see that you have already implemented this in this fashion for the managedinstance .. So it is fine if you want to keep this consistent but consider if this is something you would want in the future.
> 
> Refers to: specification/datamigration/resource-manager/Microsoft.DataMigration/preview/2022-03-30-preview/sqlmigration.json:40 in [7eb26c8](https://github.com/Azure/azure-rest-api-specs/commit/7eb26c84f50274e95658c4dfc02ac0a15c40f78b). [](commit_id = [7eb26c8](https://github.com/Azure/azure-rest-api-specs/commit/7eb26c84f50274e95658c4dfc02ac0a15c40f78b), deletion_comment = False)

In our backend we do have a distinction between migrations and migration configuration, there is only one entity object that uniquely identifies the migration resource and, each time it generates new FSM operations, but there is only one underlying migration entity object.

File: specification/datamigration/resource-manager/Microsoft.DataMigration/preview/2022-03-30-preview/examples/SqlDbCreateOrUpdateDatabaseMigrationMAX.json (hunk 1 above)
Comment (ashutoshsuman99) [reply specification/datamigration/resource-manager/Microsoft.DataMigration/preview/2022-03-30-preview/examples/SqlDbCreateOrUpdateDatabaseMigrationMAX.json:36]: done

**Comment by ashutoshsuman99** [comment]
> > ```
> >         "description": "The child resources to include in the response.",
> > ```
> 
> Update description.
> 
> Refers to: specification/datamigration/resource-manager/Microsoft.DataMigration/preview/2022-03-30-preview/sqlmigration.json:50 in [7eb26c8](https://github.com/Azure/azure-rest-api-specs/commit/7eb26c84f50274e95658c4dfc02ac0a15c40f78b). [](commit_id = [7eb26c8](https://github.com/Azure/azure-rest-api-specs/commit/7eb26c84f50274e95658c4dfc02ac0a15c40f78b), deletion_comment = False)

Done

**Comment by raosuhas** [comment]
> > > ```
> > >     },
> > > ```
> > 
> > 
> > Why is the scope duplicated here ? It seems like since this is an extension resource the scope is defined by the resource itself ?
> > Refers to: specification/datamigration/resource-manager/Microsoft.DataMigration/preview/2022-03-30-preview/sqlmigration.json:1262 in [7eb26c8](https://github.com/Azure/azure-rest-api-specs/commit/7eb26c84f50274e95658c4dfc02ac0a15c40f78b). [](commit_id = [7eb26c8](https://github.com/Azure/azure-rest-api-specs/commit/7eb26c84f50274e95658c4dfc02ac0a15c40f78b), deletion_comment = False)
> 
> Yes, this is a proxy resource. Scope in the request body is to validate permission. When you send request, arm first validates that user has permission on MI/VM/SQLDB. Arm can validate permission on property in input, so we have specified 'scope' in manifest as the property to be validated.

You dont need to do this since the scope permission is implicit since it is an extension resource. User will not be able to do add the extension resource unless they have permissions on the base resource. Are you looking for some higher level of permissions here ?

**Comment by hitenjava** [comment]
> You dont need to do this since the scope permission is implicit since it is an extension resource. User will not be able to do add the extension resource unless they have permissions on the base resource. Are you looking for some higher level of permissions here ?

Yes, we are looking for write permission on the resource. This is what we have in manifest:
image

**Comment by raosuhas** [comment]
Had offline discussion with Ashutosh. Signing off from ARM side.

**Comment by qiaozha** [comment]
@raych1 Could you take a look at the model validation error, it looks like a false alarm.

**Review by qiaozha — APPROVED** [review]

**Pull request merged**